### ISSUE TITLE
DISCUSSION 3: `subscribeAndSpyOn` + `unsubscribe` VS `spyOn` and `dispose`

### ISSUE THREAD

**Issue by shairez** [hirezio/observer-spy#15]
### CONTEXT:
Following #12 (by the amazing @ThomasBurleson) we had a discussion and realized this PR could be separated into several PRs, and before that we need to discuss them. 

**This one is about:** changing the factory method and functionality of `ObserverSpyWithSubscription`.

## The suggested API changes to `ObserverSpy` - 

```ts

export type SpyOnResults<T> = [ObserverSpy<T>, () => void, Observable<any>];

export function spyOn<T>(
  source$: Observable<T>,
  completionCallback?: CompletionCallback
): SpyOnResults<T> {
  const spy = new ObserverSpy<T>(completionCallback);
  const subscription = source$.subscribe(spy);
  const dispose = () => {
    subscription.unsubscribe();
    removeSpyFromWatchList(spy);
  };
  subscriptions.push([spy, dispose, source$]);

  return [spy, dispose, source$];
}

```

## Thomas' points:

- `spyOn` is shorter and cleaner
- `dispose()` relates to the spy, not the subscription. (you shouldn't care about what happens behind the scenes)
- tuples are better because they let you name your own variables during the array destructuring.
- You separate responsibilities and leave the spy to handle spying alone. 

## Shai's points: 

- just the name `spyOn` hides the fact that there is a subscription happening behind the scenes, and because the subscription IS the action of the test, it's an important information to convey. `subscribeAndSpyOn` shows what actually happens behind the scenes better IMO. 
- Because of jasmine's `spyOn` function, when scanning the test, people might confuse the two and not realize there's a subscription involved here. 
- I always prefer to return properties over tuples, especially if I want to keep things consistent between lots of tests. It also helps to understand the API better IMO. 
- I prefer the name `unsubscribe` over `dispose` because it's a term + functionality people already know. 

---

I'd love to hear your thoughts / corrections @ThomasBurleson @katharinakoal @edbzn @burkybang @yjaaidi (and whoever is interested to share their opinion), before we decide whether to break it into its own PR. 

### **SO... WHAT DO YOU THINK?**

**Comment by NetanelBasal** [comment]
Hey Guys, great work! Here are my points :)

- I prefer the `subscribeAndSpyOn` over `new ObserverSpy()` call. I think its functionality is what most users needs - subscribe and spy. I think the name should be shorter, something like `spyOnObserver` or `subscribeSpyTo`.

- I prefer to return an object rather than an array for two reasons: The order doesn't matter and name consistency.
- I like the additional functionality of automatically unsubscribe inside the `afterEach` call.
- I think that in this case, `unsubscribe` is more accurate than `dispose`.

Last but not least, the `README` should indicate that this library is meant for advanced use cases. If users need to write simple specs, they could create the following function:

```ts
function spyOnObserver<T>(source: Observable<T>, baseName = '') {
  const spy = jasmine.createSpyObj<Observer<T>>(baseName, [
    'next',
    'error',
    'complete',
  ]);

  return {
    unsubscribe: source.subscribe(spy),
    spy,
  };
}
```

Great work @shairez @katharinakoal @ThomasBurleson

**Comment by yjaaidi** [comment]
🤔 very interesting!

I agree with **`unsubscribe`** over `dispose` and **props** over tuples.

Concerning `new ObserverSpy()` vs a factory function, I wonder if `ObserverSpy` shouldn't be internal to avoid using the same `ObservableSpy` twice. Using a factory function user cannot dissociate spy creation and subscription.

What about:
```typescript
const spy = ObservableTestBed.subscribeTo(source$);

spy.unsubscribe();

// and

afterEach(() => ObservableTestBed.unsubscribe())

// and

afterEach(() => ObservableTestBed.verifyAllComplete()); // or something like that to check that all observables are complete
```

**Comment by shairez** [comment]
Thanks (again) for the feedback @NetanelBasal and @yjaaidi ! ❤️ 

Netanel - I like the name `subscribeSpyTo` better than `spyOnObserver` because it describes the important thing IMO -the subscription.  

@katharinakoal - what do you think about us changing `subscribeAndSpyOn` to `subscribeSpyTo` ?

I also agree that we should change the README accordingly, I'm just waiting for this issues to be resolved first. 

Younes - 
Regarding making not subscribing twice.. well.. what prevents them from manually subscribing the `spy` returned from the factory function? 😄 

Regarding the `verifyAllComplete` - I like the idea.
I don't think `ObservableTestBed` is necessary (unless it was an instance, and then you could "scope" spies per spec file and aggregate them per spec file and not ALL the files).

factory functions (without an object) feel cleaner (to me). 

Maybe something like:

```ts
import {ObserverSpyFactories} from '@hirez_io/observer-spy`;

const {subscribeSpyTo, cleanSubscriptions}  = ObserverSpyFactories();

```

Then you can preserve the scope (and minimize leakage if someone won't call "unsubscribe" or "clean subscription".


BUT...

maybe this could be handled globally and throwing if you're leaking somewhere... 🤔

**Comment by yjaaidi** [comment]
Concerning the naming of the function, it would be great to find a nice metaphor where one verb could be enough... but nothing comes to my mind right now. I thought about `track`, `monitor`, `watch`... but none implies that the observable has been subscribed to. `subscribeSpyTo` sounds like the best fit right now.

>Younes -
Regarding making not subscribing twice.. well.. what prevents them from manually subscribing the spy returned from the factory function? 😄

Oh! My bad. I didn't mention that this would come with the deprecation of the `subscribeTo` method. 😊

I totally agree that factory functions look better and give more control and less surprises but it needs more boilerplate if you want to use it in `afterEach` for example:
```typescript
let subscribeSpyTo;
let cleanSubscriptions;
beforeEach(() => ({subscribeSpyTo, cleanSubscriptions} = ...));
afterEach(() => cleanSubscriptions());
```

vs. 

```typescript
afterEach(() => ObservableTestBed.verifyAllComplete());
```

It's not that bad to use global stuff if it makes tests easier to read & write. Tests will run sequentially and if they run in parallel they will be isolated by file in different scopes _(except probably when using Jest's experimental `test.concurrent` and that would be an edge case I wouldn't care about right now... if the problem appears, we could provide a `TestBed` factory function)_

**Comment by katharinakoal** [comment]
Naming this little function is quite an interesting problem!
I also think `subscribeSpyTo` is nicer than `subscribeAndSpyOn`. 

Additionally, this issue got me thinking about the central concept of spies (at least in terms of testing 🕵️). In my understanding, if you spy on an object or a method, you usually track its interaction with other parts of your system and optionally mock its behavior. But an ObserverSpy behaves more like an enhanced observer rather than a spy  (IMO). That's why I think a plain `subscribeTo()` might be sufficient. Or maybe just `observe()`? Still thinking...

**Comment by yjaaidi** [comment]
Great point @katharinakoal!
In fact, it’s a spy that always calls through and we are doing nothing except subscribing and observing. The only problem with verbs like monitor, track or observe is that it’s not explicitly saying that it will subscribe immediately.

I feel like if it's a function, `subscribeTo` would be too generic, and we would have to check the imports to understand that it comes from `observer-spy`. If it comes from an object or class with static methods like `ObserverTestBed.subscribeTo()` then that would be explicit enough probably.

**Comment by shairez** [comment]
Sorry for the late response friends, had some sad family related issues to attend to. 

@yjaaidi - 
Regarding using an object vs a factory function... 

I meant to do it per file, not per `beforeEach`. 
But even if you do it globablly - you get the same result as with an object but with less code - 

```ts
import {subscribeSpyTo, cleanSubscriptions} from '@hirez_io/observer-spy';

it('should...', ()=>{
  subscribeSpyTo()
});

afterEach(()=>{
  cleanSubscriptions();
});
```

No? (or am I missing something...)

@katharinakoal - 
Good point, though I think of spies as anything you can "interrogate" after it's been called. 
That's why I see the observer spies and function spies to be similar: you can still call the functions, but you have state that tracks it's usage... but because observers are push based, they are more passively than functions and that's why the have less stuff to track about them. 

I agree with Younes that `subscribeTo` or `observe` are still too general and might still confuse the reader (and might also potentially overlap with other third party deps in the future).

So what do you we stick to `subscribeSpyTo` and make that change?

**Comment by edbzn** [comment]
Hi there, I'm in favor of deprecating `subscribeAndSpyOn`, renaming to `subscribeSpyTo` and keeping the `unsubscribe` name as it.

**Comment by katharinakoal** [comment]
Sounds good to me, too! 👌
